src/pool/poolClient.ts

```diff
@@ -217,9 +219,56 @@ export class PoolEventState {
 }
 
 class UserState {
-  constructor(private contract: uma.clients.erc20.Instance) {}
-  public async read(user: string) {
+  private seen = new Set<string>();
+  private events: uma.clients.erc20.Transfer[] = [];
+  constructor(private contract: uma.clients.erc20.Instance, private startBlock = 0) {}
+  private makeId(params: EventIdParams) {
+    return [params.blockNumber, params.transactionIndex, params.logIndex].join("!");
+  }
+  hasEvent(params: EventIdParams) {
+    return this.seen.has(this.makeId(params));
+  }
+  private addEvent(params: EventIdParams) {
+    return this.seen.add(this.makeId(params));
+  }
+  private filterSeen = (params: EventIdParams) => {
+    const seen = this.hasEvent(params);
+    if (!seen) this.addEvent(params);
+    return !seen;
+  };
+  public async readEvents(user:string, endBlock: number) {
+    if (endBlock <= this.startBlock) return [];
+    const events = (
+      await Promise.all([
+        ...(await this.contract.queryFilter(this.contract.filters.Transfer(user,undefined), this.startBlock, endBlock)),
+        ...(await this.contract.queryFilter(this.contract.filters.Transfer(undefined,user), this.startBlock, endBlock)),
+      ])
+    )
+    .filter(this.filterSeen)
+    // filter out mint/burn transfers
+    .filter((event:uma.clients.erc20.Transfer) => event.args.from !== AddressZero && event.args.to !== AddressZero)
+    .flat()
+
+    this.events = this.events
+      .concat(events)
+      .sort((a, b) => {
+        if (a.blockNumber !== b.blockNumber) return a.blockNumber - b.blockNumber;
+        if (a.transactionIndex !== b.transactionIndex) return a.transactionIndex - b.transactionIndex;
+        if (a.logIndex !== b.logIndex) return a.logIndex - b.logIndex;
+        // if everything is the same, return a, ie maintain order of array
+        return -1;
+      });
+    // ethers queries are inclusive [start,end] unless start === end, then exclusive (start,end). we increment to make sure we dont see same event twice
+    this.startBlock = endBlock + 1;
+    return this.events
+  }
+  public async read(user: string, endBlock:number) {
```

Comment:
What's the difference between read and readEvents? Just from the interface, it's somewhat unclear what these methods are doing and why they're different, since they have the same args, but a different name.